### ISSUE TITLE
Disallow check for thruthy enum values

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#7473]
I think I have been bitten enough times of checking of a truthy enum value
```ts
const enum Char {
    A,
    B,
    C,
}

let char: Char;
if (Date.now() % 2 > 0) {
   char = Char.A;
}
if (char) {
   // do something
}
```

The above has a bug. And that is `char` can have a `0` value. And that is a falsy value. I think there should be an error here and people should write:

```ts
if (typeof char !== 'undefined') {
   // do something
}
```

or any of the other non-truthy checks.

I don't think people would ever want to check for truthy on an enum value(which can have a `0` value). Which makes putting it behind a flag less relevant.

**Comment by tinganho** [comment]
Maybe also consider all types of number values.

**Comment by mhegazy** [comment]
looks like the same as https://github.com/Microsoft/TypeScript/issues/7405

**Comment by tinganho** [comment]
Yes I saw that. But that suggest enum values to start with 1?

**Comment by mhegazy** [comment]
So what is the proposal?

**Comment by tinganho** [comment]
This proposal is for checking. It should throw an error when you check for thruthy value of an enum value that can take the value 0.

**Comment by aluanhaddad** [comment]
Just my 2 cents, but I don't think this is an emun issue, or even a TypeScript issue. TypeScript's semantic notion of enums is standard across many languages in the same family. 

Here is my rational:
1. Enums starting at 0 by default is what people expect.
2. Enums are, semantically, just named constant numbers, great improving readability.
3. JavaScript treats 0 as falsy, so any falsy check likely to break when the checked value is a number. This is consistent with JavaScript.
4. In languages like C implicit conversions in conditionals have the same hazards.
5. Even in C# which, provides does not provide falsy checking for values besides bools (unless you define an operator to imbue such semantics into your type), has a hazard around enums because they are value types. That is one reason why it is considered good practice in C# to define an explicit `Unspecified = 0` or `None = 0` in enums.

Now, all that said, the art of language design is always moving forward, and the TypeScript and C# teams are a great example of this, so the issue of conflating 0 with its corresponding enum value is something that may be solved in the future. But I think the fundamental problem is not in enums but in the naive use of falsy checking.

I suggest that, as a general practice, enums should always be defined with the 0 value being equivalent to the "Zero Like", "Unspecified", or "Falsy" value for that enum type.
I have found that sticking to this approach has made code cleaner and more resilient in a number of languages.

**Comment by tinganho** [comment]
@aluanhaddad I'm not suggesting enum values to start with 1.

**Comment by tinganho** [comment]
Though I'm not convinced starting with zero is optimal. My argument is that an enumeration should have the same underlying type. And 0 for me is a slightly different type than 1, 2, 3 etc. It is the null for numbers.

**Comment by aluanhaddad** [comment]
> And 0 for me is a slightly different type than 1, 2, 3 etc. It is the null for numbers.

@tinganho that is exactly my point, and why I suggested you define an explicit additional value corresponding to the `0` of the enum. That way you avoid the problem. For example. Suppose you are deserializing an object which is represented using a parallel enum type on the server side. Using `JSON.parse`, depending on the backend api's implementation, you wont know whether the value was set to 0 or if it was default initialized to 0. 

That said, I was speaking very broadly, and coming from a philosophical place.



> I'm not suggesting enum values to start with 1.... Though I'm not convinced starting with zero is optimal.

Wouldn't it be a major breaking change to alter this behavior?

**Comment by DanielRosenwasser** [comment]
Currently we have enums in the compiler that utilize falsy/truthy behavior (i.e. `Ternary`), and it's a reasonable pattern to have an enum that functions as a boolean, but gives explicit name. For instance,

```ts
enum ErrorHandling {
    IgnoreErrors,
    ReportErrors
}
```

**Comment by aluanhaddad** [comment]
My feeling is that changes would break existing code, and offer little to no benefit.

**Comment by tinganho** [comment]
@DanielRosenwasser your example makes sense with checking for thruthy. So I'm closing because it will break existing code.

>My feeling is that changes would break existing code, and offer little to no benefit.

Maybe it is just me, but I'm pretty blind when it comes to finding the error for thruthy checks. Because you expect it to be truthy. The 0 value is very hidden and have made me lost countless hours.

**Comment by RyanCavanaugh** [comment]
As usual I'd suggest someone write a TSLint rule :wink:
